Add coordinator to vesync

iprak opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This request switches Vesync devices to use DataCoordinator pattern instead of each entity updating itself. The latter can run over the daily webcalls limits where the web end point returns an error message like `Current user's daily request quota has been used up. The quota formula is \"3200 + 1500 * user owned device number\". Please note that we may change the constant in the formula.`


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

This should partially address #131781 which in my case was due to daily quota being used up.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #131781
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

iprak committed:
Using DataCoordinator instead of direct update

Using DataCoordinator allows us to reduce number
of web calls being made. Some devices can have
multiple entities which increases the number of
web calls and hitting the daily limit.
DataCoordinator reduces that to just 1 call.

joostlek commented on homeassistant/components/vesync/__init__.py:
```diff
@@ -48,6 +51,25 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> b
     hass.data[DOMAIN] = {}
     hass.data[DOMAIN][VS_MANAGER] = manager
 
+    # Create a DataUpdateCoordinator for the manager
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        try:
+            await hass.async_add_executor_job(manager.update)
+        except Exception as err:
+            raise UpdateFailed(f"Update failed: {err}") from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=DOMAIN,
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=30),
+    )
```
Let's make this a separate class in `coordinator.py`. A lot of other integrations have a `coordinator.py` and are good examples for this

cdnninja commented on homeassistant/components/vesync/__init__.py:
```diff
@@ -48,6 +51,25 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> b
+        _LOGGER,
+        name=DOMAIN,
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=30),
```
Could we set this to 60? At 30 it would max the count at 3 devices.   60 will solve that.

iprak replied:
It should not max out, there should be only one request regardless of the number of devices per second which would be (25 * 3600 / 30) = 2880 per day, right?

The error message mentioned "3200 + 1500 * user owned device number" which was a issue if there were lots of total entities.

cdnninja replied:
Ahh let me check the underlying code. I assumed that Vesync library was just doing a for each loop.  The endpoints are different so I believe it is but I'll double check.

iprak replied:
This [update](https://github.com/webdjoe/pyvesync/blob/c44011abd61b7801f90264cee10676ceb77285a2/src/pyvesync/vesync.py#L418) on the vesync manager updates all the devices within a 30 sec interval check.

webdjoe replied:
There will be another 2880 calls per day due to pulling the device list in each `manager.update()` call, which goes beyond the limit. It would be better to use `manager.update_all_devices()` to avoid the unnecessary pulling of the device list. Adding / removing devices can be done manually by a service

iprak replied:
Oh I see there is call to `get_devices`. That makes sense `update_all_devices` calls `dev.get_details` and the same is called from `update`.

This should get the count back to 2800 + a couple for HA startup. I will try to wrap all this tonight.

iprak replied:
We can't use `update_all_devices` because that method is incorrectly iterates over `self._dev_list.keys` instead of `self._dev_list.values`. 

I can continue using update for now and then update it one vesync has been fixed?

webdjoe replied:
I can't believe I never caught that. 

@joostlek can I make a release that can be bumped in this PR?

webdjoe replied:
2.1.15 is published with the fix

iprak replied:
Thanks for the prompt fix. I rebase my branch and bumped the pevsync version and also selected the `Dependency upgrade` checkbox in the PR.

gjohansson-ST replied:
Could we include the bump of the lib in a preliminary PR so it's not included in this (quite large already) PR

iprak replied:
Sure. I have submitted https://github.com/home-assistant/core/pull/134156 for consuming 2.1.15. I will adjust this PR once that has been merged.

cdnninja replied:
On the topic of the calls in the above call it does the following: 

`            for device in chain(*devices):
                device.update()`
https://github.com/webdjoe/pyvesync/blob/c44011abd61b7801f90264cee10676ceb77285a2/src/pyvesync/vesync.py#L435-L436

Which calls once for each device.  Based on this I still believe you will max out calls if not doing 60 second updates.
30 second interval will be: 
1 device = 2880 update requests,  4700 quota
2 devices = 5760 update requests, 6200 quota
3 devices = 8640 update requests, 7700 quota. 

This doesn't account for any room on device list updates, HA restarts etc.     We could go lower than 60 if wanting to consume some of the base 3200 but I was thinking use the "device" added qty and leave the 3200 for restarts, mobile app usage etc.  As such 1500 calls per device is 1.04 calls per minute.

iprak replied:
I see your point. I think I was more fixated on the issue that each sensor entity was also doing its update. So this PR should still help event for outlet which have energy sensors.

But at device level, 60sec should work. 1500 request works out to be 57.6 seconds.

We can separately consider making this a configuration or dynamically adjusting the next refresh based on device count.

Thank you for your input.

cdnninja replied:
I hoped to go dynamic but not aloud by HA. https://github.com/home-assistant/core/pull/127152

iprak replied:
Bummer ..  then we can adjust the `update_interval` of the coordinator.

iprak replied:
@gjohansson-ST Would you know if I need to do something else for https://github.com/home-assistant/core/pull/134156 ? Does it need additional approvals?

gjohansson-ST replied:
Please rebase this PR to include the latest lib bump. Setting this to draft for now, please click "ready for review" when it is.

iprak committed:
Moved data coordinator to separate class

iprak committed:
Using update_all_devices instead of update

This avoids fetching device list every time.

iprak committed:
Created separate test data for air-purifier-400s

air-purifier-400s-detail-updated.json is a copy of
device-details.json but should allow us to test
fan devices separately.

iprak committed:
Added test for fan device update

The air-purifier-400s device is updated via
coordinator and then cross checked for the
`night_light` attribute.

iprak committed:
Increased update interval to 60s

joostlek commented on homeassistant/components/vesync/__init__.py:
```diff
@@ -48,6 +50,11 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> b
 
+    coordinator = VeSyncDataCoordinator(hass, manager)
+
+    # Store coordinator at domain level since only single integration instance is permitted.
```
I think it would be nice if we could open up the integration to allow multiple instances

iprak replied:
I think implementing that would be outside the scope of this PR. I think multiple integrations would only be possible if multiple logins are supported.

cdnninja replied:
Migrating location of coordinator later would add more work so it is a fair point.

joostlek replied:
Oh definitely different PR, but I still wanted to mention it

joostlek commented on homeassistant/components/vesync/const.py:
```diff
@@ -4,10 +4,25 @@
 SERVICE_UPDATE_DEVS = "update_devices"
 
+UPDATE_INTERVAL = 60
+"""
+Update interval for DataCoordinator.
+
+The vesync daily quota formula is 3200 + 1500 * device_count.
+
+An interval of 60 seconds amounts 1440 calls/day which
+would be below the 4700 daily quota. For 2 devices, the
+total would be 2880.
+
+Using 30 seconds interval gives 8640 for 3 devices which
+exceeds the quota of 7700.
+"""
```
Why don't we just make this dynamic? We know how many devices someone has so we can calculate how many API calls we can make.

But this would be for a followup

iprak replied:
Sounds good

joostlek commented on homeassistant/components/vesync/coordinator.py:
```diff
@@ -0,0 +1,42 @@
+"""Representation of VeSync data coordinator."""
```
It's not a representation

joostlek commented on homeassistant/components/vesync/coordinator.py:
```diff
@@ -0,0 +1,42 @@
+            return await self.hass.async_add_executor_job(
+                self._manager.update_all_devices
+            )
+        except Exception as error:
+            raise UpdateFailed(error) from error
```
Let's not catch bare exceptions. Please be more specific with what can be raised

iprak replied:
The call into pyvesync is from device.update > [device.get_details](https://github.com/webdjoe/pyvesync/blob/736c29256a5a8b562ddb1b16ac20e418b3e27f29/src/pyvesync/vesyncfan.py#L387) > Helpers.call_api.

pyvesync catches all exceptions and does not rethrow them. It also performs `get` on dictionaries safely so I don't think we will get an exception. If data was malformed or request failed, then we will just not get any data. I am removing the catch.

joostlek commented on homeassistant/components/vesync/coordinator.py:
```diff
@@ -0,0 +1,42 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class VeSyncDataCoordinator(DataUpdateCoordinator):
```
```suggestion
class VeSyncDataCoordinator(DataUpdateCoordinator[..]):
```
Add the return type of `_async_update_data` to the type slot of the coordinator

joostlek commented on homeassistant/components/vesync/entity.py:
```diff
@@ -6,17 +6,24 @@
 
 
-class VeSyncBaseEntity(Entity):
+class VeSyncBaseEntity(CoordinatorEntity, Entity):
```
```suggestion
class VeSyncBaseEntity(CoordinatorEntity[..]):
```
Add the coordinator type to the typeslot of the coordinator entity

joostlek commented on homeassistant/components/vesync/entity.py:
```diff
@@ -6,17 +6,24 @@
 
-    def __init__(self, device: VeSyncBaseDevice) -> None:
+    def __init__(
+        self, device: VeSyncBaseDevice, coordinator: DataUpdateCoordinator
```
Let's just use the implemented coordinator instead

joostlek commented on homeassistant/components/vesync/entity.py:
```diff
@@ -46,9 +53,12 @@ def device_info(self) -> DeviceInfo:
-    def update(self) -> None:
-        """Update vesync device."""
-        self.device.update()
+    async def async_added_to_hass(self) -> None:
+        """When entity is added to hass."""
+        await super().async_added_to_hass()
+        self.async_on_remove(
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
```
Why do we do this?

iprak replied:
I don't think we need to take any additional action for this integration on remove. I have removed it.

joostlek commented on tests/components/vesync/test_fan.py:
```diff
@@ -32,6 +37,9 @@ async def test_fan_state(
     await hass.config_entries.async_setup(config_entry.entry_id)
     await hass.async_block_till_done()
 
+    coordinator = hass.data[DOMAIN][VS_COORDINATOR]
+    assert coordinator
```
Please don't touch internals

iprak replied:
I was following patterns implemented in other integrations where data in hass.data[DOMAIN] is accessed and even this integration was already doing something like that [here](https://github.com/home-assistant/core/blob/55dc4b0d2c7450955e8d0cb37a0031aee61bed9b/tests/components/vesync/test_init.py#L57).

joostlek replied:
Yep, we have quite some occurences in the code base, but we are trying to get them out slowly. So let's not try to add new ones

joostlek commented on tests/components/vesync/test_fan.py:
```diff
@@ -49,3 +57,15 @@ async def test_fan_state(
     for entity in entities:
         assert hass.states.get(entity.entity_id) == snapshot(name=entity.entity_id)
+
+    # Test for update via coordinator using entity air_purifier_400s
+    # Update the request mock to return different data
+    mock_air_purifier_400s_update_response(requests_mock)
+    await coordinator.async_refresh()
+    await hass.async_block_till_done()
+
+    for entity in entities:
+        if entity.entity_id == "fan.air_purifier_400s":
+            assert hass.states.get(entity.entity_id) == snapshot(
+                name=f"{entity.entity_id}_updated"
+            )
```
Instead of using snapshot tests here, let's just use normal tests here as we want to test something changed, and with snapshots you're hiding that

iprak replied:
I would love to do that but I am not really sure how to proceed. I was attempting to test data fetch update e.g. water level changed or something was changed on the device physically and not through the integration.

joostlek replied:
Yes the idea is to repatch the library and return what you expect to return and then use the freezer to skip time to trigger a natural update (initiated by the integration code itself)

iprak replied:
Ok that is what I too was thinking. I will try to create a separate test just for the coordinator part. All the existing tests in this integration are based on snapshots.

joostlek replied:
But let's not use snapshots for this use case, as they are great for fixating big data objects, not for testing change

iprak replied:
I am running into some issues with the testing changes in my last commit and can use your help.

Below is the debug log when running pytest. 

The values like `400s-purifier-filter-life` are being printed in `native_value` getter. The line `air_purifier_400s_air_quality=5` was printed just before the final assert in the test.
The output indicates that `native_value` was invoked but after the assert and it does returns the correct value. It seems than I am missing something.

```
INFO:homeassistant.components.sensor:Setting up vesync.sensor
INFO:homeassistant.helpers.entity_registry:Registered new sensor.vesync entity: sensor.air_purifier_400s_filter_lifetime
400s-purifier-filter-life=99
INFO:homeassistant.helpers.entity_registry:Registered new sensor.vesync entity: sensor.air_purifier_400s_air_quality
400s-purifier-air-quality=5
INFO:homeassistant.helpers.entity_registry:Registered new sensor.vesync entity: sensor.air_purifier_400s_pm2_5
400s-purifier-pm25=1
INFO:homeassistant.helpers.entity_registry:Registered new fan.vesync entity: fan.air_purifier_400s
WARNING:asyncio:Executing <Task pending name='Task-3' coro=<test_coordinator_update() running at /workspaces/core-fork/tests/components/vesync/test_coordinator.py:60> cb=[_run_until_complete_cb() at /usr/local/lib/python3.13/asyncio/base_events.py:181] created at /usr/local/lib/python3.13/asyncio/tasks.py:748> took 120.000 seconds

air_purifier_400s_air_quality=5

400s-purifier-filter-life=99
400s-purifier-air-quality=15
400s-purifier-pm25=1
```

iprak replied:
Forcing hass to wait for background jobs to finish `await hass.async_block_till_done(True)` does address the sequence. Is that the correct thing to do?

iprak committed:
Removed exception catch and updated description

iprak committed:
Added type slot for coordinator

iprak committed:
Using VeSyncDataCoordinator instead of DataUpdateCoordinator

iprak committed:
Removed async_added_to_hass

iprak committed:
Reverted snapshot testing for fan

iprak committed:
Created separate tests for coordinator

iprak committed:
Fixed the failing test for coordinator

cdnninja commented on homeassistant/components/vesync/coordinator.py:
```diff
@@ -0,0 +1,37 @@
+        """Fetch data from API endpoint."""
+
+        # Using `update_all_devices` instead of `update` to avoid fetching device list every time.
+        return await self.hass.async_add_executor_job(self._manager.update_all_devices)
```
Do you know if this call also update energy for plugs?  Since I think that is a separate call? Should entity update also be removed from switch.py?

iprak replied:
I assumed that `update` took care of that but that doesn't seem to be the case for outlets. Do you know why it might been designed as a separate thing? I can invoke a second job for `VeSync.update_energy` to handle that though. Thoughts?

cdnninja replied:
Yes I think just calling update energy should do it!

iprak replied:
Okay fixed that too. I ended up writing more unit tests. The tests can be improved, there is some duplicate content but I think it would be okay to do that in separate PR. My tests kept failing till I discovered that energy updates are only permitted every 6 hours.

iprak committed:
Update energy as well

iprak committed:
Added tests for energy update

The new file defines the data for the "Outlet"
device.

cdnninja reviewed (changes requested):
I would think this section should be deleted: 

https://github.com/iprak/core-fork/blob/8b7e5777dd8430860e6b441362d7a412c809d59a/homeassistant/components/vesync/switch.py#L82-L85

joostlek commented on homeassistant/components/vesync/coordinator.py:
```diff
@@ -0,0 +1,39 @@
+            update_interval=timedelta(seconds=UPDATE_INTERVAL),
+        )
+
+    async def _async_update_data(self) -> dict[str, Any]:
+        """Fetch data from API endpoint."""
+
+        # Using `update_all_devices` instead of `update` to avoid fetching device list every time.
+        await self.hass.async_add_executor_job(self._manager.update_all_devices)
+        # Vesync updates energy on applicable devices every 6 hours
+        return await self.hass.async_add_executor_job(self._manager.update_energy)
```
Changing from event loop to an executor job is quite expensive. So instead of running 2 after each other, create 1 function and run that in the executor

joostlek commented on tests/components/vesync/test_coordinator.py:
```diff
@@ -0,0 +1,92 @@
+
+    # Test energy update
+    # pyvesync only updates energy parameters once every 6 hours and hence the 21600.
+    freezer.tick(timedelta(seconds=21600))
```
Please just set it to 6 hours

joostlek commented on tests/components/vesync/test_coordinator.py:
```diff
@@ -0,0 +1,92 @@
+from tests.common import MockConfigEntry, async_fire_time_changed
+
+
+async def test_coordinator_update(
```
This is rather testing behavior of the entities, so let's move this to a platform test file instead

iprak replied:
Renamed the file and test

iprak committed:
Fixed typing and executor callback

iprak committed:
Removed unused VeSyncSwitchHA.update

iprak committed:
Renamed to platform

gjohansson-ST commented on homeassistant/components/vesync/coordinator.py:
```diff
@@ -34,4 +34,6 @@ async def _async_update_data(self) -> dict[str, Any]:
-        return await self.hass.async_add_executor_job(self._manager.update_all_devices)
+        await self.hass.async_add_executor_job(self._manager.update_all_devices)
+        # Vesync updates energy on applicable devices every 6 hours
+        return await self.hass.async_add_executor_job(self._manager.update_energy)
```
Put both calls in a new function and use that function as an argument for the executor job

iprak replied:
That has been addressed as part of this review- https://github.com/home-assistant/core/pull/134087#pullrequestreview-2527205653

gjohansson-ST replied:
Sorry, clearly looked at an old commit. 👍
